services/metric/registry.go

```diff
@@ -0,0 +1,232 @@
+	simpleEwmas sync.Pool
+	varEwmas    sync.Pool
+	sets        sync.Pool
+}
+
+func (r *registry) GetCounter(m Measurement) (Counter, error) {
+	newCounter := r.counters.Get()
+	res, putBack := r.store.LoadOrStore(m, newCounter)
+	if putBack {
+		r.counters.Put(newCounter)
+	} else {
+		r.updateIndex(m, res)
+	}
+	c, ok := res.(Counter)
+	if !ok {
+		return nil, fmt.Errorf("a different type of metric exists in the registry with the same key [%+v]: %T", m, res)
+	}
+	return c, nil
+}
+
+func (r *registry) MustGetCounter(m Measurement) Counter {
+	c, err := r.GetCounter(m)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (r *registry) GetGauge(m Measurement) (Gauge, error) {
+	newGauge := r.gauges.Get()
+	res, putBack := r.store.LoadOrStore(m, newGauge)
+	if putBack {
+		r.gauges.Put(newGauge)
+	} else {
+		r.updateIndex(m, res)
+	}
+	g, ok := res.(Gauge)
+	if !ok {
+		return nil, fmt.Errorf("a different type of metric exists in the registry with the same key [%+v]: %T", m, res)
+	}
+	return g, nil
+}
+
+func (r *registry) MustGetGauge(m Measurement) Gauge {
+	c, err := r.GetGauge(m)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (r *registry) GetSimpleMovingAvg(m Measurement) (MovingAverage, error) {
+	newEwma := r.simpleEwmas.Get()
+	res, putBack := r.store.LoadOrStore(m, newEwma)
+	if putBack {
+		r.simpleEwmas.Put(newEwma)
+	} else {
+		r.updateIndex(m, res)
+	}
+	ma, ok := res.(MovingAverage)
+	if !ok {
+		return nil, fmt.Errorf("a different type of metric exists in the registry with the same key [%+v]: %T", m, res)
+	}
+	return ma, nil
+}
+
+func (r *registry) MustGetSimpleMovingAvg(m Measurement) MovingAverage {
+	ma, err := r.GetSimpleMovingAvg(m)
+	if err != nil {
+		panic(err)
+	}
+	return ma
+}
+
+func (r *registry) GetVarMovingAvg(m Measurement, age float64) (MovingAverage, error) {
+	newEwma := r.varEwmas.Get()
+	decay := 2 / (age + 1)
+	newEwma.(*VariableEWMA).decay = decay
+	res, putBack := r.store.LoadOrStore(m, newEwma)
+	if putBack {
+		r.varEwmas.Put(newEwma)
+	} else {
+		r.updateIndex(m, res)
+	}
+	ma, ok := res.(*VariableEWMA)
+	if !ok {
+		return nil, fmt.Errorf("a different type of metric exists in the registry with the same key [%+v]: %T", m, res)
+	}
+	if ma.decay != decay {
+		currentAge := 2/ma.decay + 1
+		return nil, fmt.Errorf("another moving average with age %f instead of %f exists in the registry with the same key [%+v]: %T", currentAge, age, m, res)
+	}
+	return ma, nil
+}
+
+func (r *registry) MustGetVarMovingAvg(m Measurement, age float64) MovingAverage {
+	ma, err := r.GetVarMovingAvg(m, age)
+	if err != nil {
+		panic(err)
+	}
+	return ma
+}
+
+func (r *registry) Range(f func(key interface{}, value interface{}) bool) {
+	r.store.Range(f)
+}
+
+func (r *registry) GetMetricsByName(name string) []TagsWithValue {
+	metricsSet, ok := r.nameIndex.Load(name)
+	if !ok {
+		return []TagsWithValue{}
+	}
+	values := []TagsWithValue{}
+	for _, value := range metricsSet.(map[Measurement]TagsWithValue) {
```

Comment:
Shouldn't we be checking the success of typecast? Like we do for the above functions

Reply:
> Shouldn't we be checking the success of typecast? Like we do for the above functions

Not really since the `Registry` has full ownership of what gets inserted in the `nameIndex` map. Only values generated by the `indexGenerator` can be added to the `nameIndex` which are of the appropriate type.
In the future, as soon as generics make their way into the `sync` package we will be able to have `sync.Map[K, V]` and this and other similar casts will become redundant :)